Remove finalizers on CSI snapshot objects

bobfuru opened the pull request:
[BZ1896084](https://bugzilla.redhat.com/show_bug.cgi?id=1896084)
Adds instructions for removing finalizers to support volume snapshot delete operations. Includes note that force option is not sufficient.

bobfuru commented:
@huffmanca and @duanwei33 PTAL for technical and QE review. Also, please confirm that this is for 4.7+ only. (CSI Snapshot is GA in 4.7.) Thanks.

huffmanca commented:
I'm very hesitant on including instructions for removing VolumeSnapshot finalizers as the **default** documentation. What I'd do is:

1. Leave the current instructions alone. These should be sufficient for the majority of use-cases.
2. Add an optional section after the existing instructions. This should say something similar to the following:

> While the above instructions satisfy the majority of delete operations, there are some circumstances where the
> VolumeSnapshot is not successfully deleted. In this case it is required to remove the existing Finalizers from the 
> VolumeSnapshot resource for the delete operation to continue....
> IMPORTANT: Only remove the Finalizers if you are confident that there are no existing references, from either 
> PersistentVolumeClaims or VolumeSnapshotContents, to the VolumeSnapshot resource.

Then in this optional section we can include the `oc patch` and `oc delete` commands.

bobfuru commented:
@huffmanca Thanks for the feedback! I reworked this so that it ends as an optional set of steps to follow if the initial default instructions are not enough. I dropped "it is required/must" wording because it falls in the "optional" step, so hopefully your suggested meaning still remains. PTAL and let me know if this ordering is logical to you.

duanwei33 commented on modules/persistent-storage-csi-snapshots-delete.adoc:
```diff
@@ -26,4 +25,46 @@ deletionPolicy: Delete <1>
+$ oc patch -n $PROJECT volumesnapshot/$NAME --type=merge -p '{"metadata": {"finalizers":null}}'
+----
+
+.. Delete the volume snapshot class:
```
@bobfuru @huffmanca I'm not sure if volumesnapshotclass deletion here is expected? Or do you want to VolumeSnapshotContent instead?

huffmanca replied:
This step shouldn't be needed at all. It can be removed.

duanwei33 commented:
@bobfuru I noticed this section name is **Deleting a volume snapshot**, so from the user side, we need a common use case and an optional part to introduce this bug fix.
In common part(I think the previous process is not complete), the **oc delete** command need to be listed to delete volumesnapshot or/and VolumeSnapshotContent, as you mentioned, if the deletionPolicy is "Delete", the VolumeSnapshotContent will be deleted alone with the volumesnapshot; if the deletionPolicy is "Retain", VolumeSnapshotContent will remain and **oc delete VolumeSnapshotContent** could be introduced here if needed.
After that, we can use the **oc patch** and **oc delete** for the exception case.

bobfuru commented:
Thanks, @duanwei33! I'm not sure that I fully captured what you are suggesting but could you PTAL and let me know if the ordering and wording make sense with my latest update?

duanwei33 commented:
> Thanks, @duanwei33! I'm not sure that I fully captured what you are suggesting but could you PTAL and let me know if the ordering and wording make sense with my latest update?

Hi @bobfuru, thanks for the update, yes this is what I mean to. Just one comment, deleting volumesnapshotcontent depends on the user requirement in the following section, so do you think the update or something like that is ok? 

Delete the volume snapshot:

$ oc delete volumesnapshot <volumesnapshot_name>
Example output
volumesnapshot.snapshot.storage.k8s.io "mysnapshot" deleted

If the deletion policy is Retain and you want to delete volumesnapshotcontent object, enter the following commands:

$ oc delete volumesnapshotcontent <volumesnapshotcontent_name>

duanwei33 commented on modules/persistent-storage-csi-snapshots-delete.adoc:
```diff
@@ -26,4 +23,65 @@ deletionPolicy: Delete <1>
 ----
 <1> If the `Delete` value is set, the underlying snapshot will be deleted, along with the `VolumeSnapshotContent` object. If the `Retain` value is set, both the underlying snapshot and `VolumeSnapshotContent` object remain.
```
A little strange for this sentence, should it be like this?
When deleting the `VolumeSnapshot` object: If the `Delete` value is set, the underlying snapshot will be deleted, along with the `VolumeSnapshotContent` object. If the `Retain` value is set, both the underlying snapshot and `VolumeSnapshotContent` object remain.

bobfuru replied:
Good point, I've added your suggestion.

bobfuru commented:
> Hi @bobfuru, thanks for the update, yes this is what I mean to. Just one comment, deleting volumesnapshotcontent depends on the user requirement in the following section, so do you think the update or something like that is ok?
> 
> Delete the volume snapshot:
> 
> $ oc delete volumesnapshot <volumesnapshot_name>
> Example output
> volumesnapshot.snapshot.storage.k8s.io "mysnapshot" deleted
> 
> If the deletion policy is Retain and you want to delete volumesnapshotcontent object, enter the following commands:
> 
> $ oc delete volumesnapshotcontent <volumesnapshotcontent_name>

It's helpful that you pointed out this out. I've made some small tweaks along the lines of your suggestion.

bobfuru commented:
@huffmanca and @duanwei33 Could you both PTAL once more and let me know if this looks good? Thanks.

duanwei33 commented:
/LGTM

bobfuru commented:
@huffmanca - Ignore my earlier comment, I think this should be backported to all versions back to 4.5, correct?

huffmanca commented on modules/persistent-storage-csi-snapshots-delete.adoc:
```diff
@@ -24,6 +21,67 @@ metadata:
+
+. Delete the volume snapshot:
+
+.. If the deletion policy is set to `Delete`, enter the following command:
```
I think we should combine the deletion of the snapshot into a single step. Then we can just include the steps to delete the VolumeSnapshotContent in the one-off case of the deletion policy being `Retain` and the user wanting to delete this content later.

bobfuru replied:
Resolved.

huffmanca commented on modules/persistent-storage-csi-snapshots-delete.adoc:
```diff
@@ -24,6 +21,67 @@ metadata:
+Even with the `--force` option, the delete operation does not delete snapshot objects until all finalizers are removed.
+====
+
+.. Remove the finalizer for the leftover resources by patching the finalizer with null:
```
While accurate, I think this would be cleaner to say:

> Remove any finalizers for the leftover resource:

We're not technically patching the finalizer, we're patching the VolumeSnapshot object to have it's finalizers removed. Your sentence is accurate, but it feels clunky to me.

bobfuru replied:
Makes sense, I rearranged this step slightly to simplify and use your wording.

huffmanca commented on modules/persistent-storage-csi-snapshots-delete.adoc:
```diff
@@ -24,6 +21,67 @@ metadata:
++
+[source,terminal]
+----
+$ oc delete volumesnapshot <volumesnapshot_name>
```
I don't believe this step is necessary if the deletion has already started. Once we execute the delete command, we get a deletionTimestamp. The controller should automatically remove any snapshots with a deletionTimestamp once the finalizers have been removed.

bobfuru replied:
Thanks, removed this redundant step.

huffmanca reviewed (commented):
Just some nitpicks and rewording. Otherwise it looks good to me!

huffmanca commented:
@bobfuru ,

In regards to backporting, VolumeSnapshots were introduced as tech preview in OCP 4.4. I think we'd need to backport it through all versions to 4.4.

bobfuru commented:
Thanks, @huffmanca!

I made a few small edits to be more in line with your most recent comments. You are correct that CSI snapshot was introduced in 4.4 but now that we're in 4.7 GA, we only support back to 4.5 as 4.4 support has ended.

bobfuru commented:
LGTM from @huffmanca via Slack.
@openshift/team-documentation PTAL

bmcelvee commented on modules/persistent-storage-csi-snapshots-delete.adoc:
```diff
@@ -24,6 +21,47 @@ metadata:
 deletionPolicy: Delete <1>
 ----
-<1> If the `Delete` value is set, the underlying snapshot will be deleted, along with the `VolumeSnapshotContent` object. If the `Retain` value is set, both the underlying snapshot and `VolumeSnapshotContent` object remain.
+<1> When deleting the volume snapshot, if the `Delete` value is set, the underlying snapshot will be deleted, along with the `VolumeSnapshotContent` object. If the `Retain` value is set, both the underlying snapshot and `VolumeSnapshotContent` object remain.
```
This is existing content and might be out of scope of this PR, but we've been avoiding future tense (e.g., will be). I'll leave it up to you if you want to change it. :)

bobfuru replied:
Good catch! I :) updated two instances using future tense in this note.

bmcelvee commented:
LGTM! Just one optional comment.

bobfuru commented:
/cherrypick enterprise-4.8

bobfuru commented:
/cherrypick enterprise-4.7

bobfuru commented:
/cherrypick enterprise-4.6

bobfuru commented:
/cherrypick enterprise-4.5
